docker/Dockerfile

```diff
@@ -1,46 +1,84 @@
+#########################
 FROM ubuntu:22.04 AS base
 
-ENV LANG en_US.utf8
-
-RUN apt-get update \
-  && apt-get -y install ca-certificates build-essential libsasl2-dev openjdk-11-jdk
-
+ARG LANG=en_US.utf8
+ENV LANG=$LANG
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN \
+    apt-get update -q -y && \
+    apt-get upgrade -q -y && \
+    apt-get install -q --no-install-recommends -y \
+        build-essential \
+        ca-certificates \
+        libsasl2-dev \
+        locales \
+        openjdk-11-jdk && \
+    rm -rf /var/lib/apt/lists/* && \
+    locale-gen ${LANG}
+
+##############################
 FROM base AS dashboard-builder
 
-RUN apt-get install -y curl gnupg protobuf-compiler && mkdir -p /etc/apt/keyrings \
-  && curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg \
-  && echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_18.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list \
-  && apt-get update && apt-get install -y nodejs
-
 COPY ./dashboard/ /risingwave/dashboard
 COPY ./proto  /risingwave/proto
 
-RUN cd /risingwave/dashboard && npm i && npm run build-static && rm -rf node_modules
-
+WORKDIR /risingwave/dashboard
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN \
```

Comment:
I think `apt-get` before `COPY` is preferred, so that the step of installing toolchain can still be cached when the code changes.

Reply:
That's fair, I'll see if I can switch it.  The main reason to move it above the `apt-get` was to be able to consolidate the `RUN` statements, which keep the image size smaller.

There is always this tradeoff between cache layers and image size.  For production uses I like to have the image size as small as possible, but for development and CI use cases I prefer to have more caching layers.

Reply:
IIUC the image size is only related with the final `risingwave` stage. `*-builder` don't need to optimize (too much) for image size.

Reply:
I moved the copy to after the apt block.